File: src/agents/queryRewriter.ts
import path from "node:path";
import fs from "node:fs";
import { type PiAgentClient } from "../piAgentClient.js";
import { allCategories, type Manifest } from "../types.js";
import { scoreText, tokenize, topTerms } from "../textUtils.js";

export class QueryRewriter {
  constructor(private knowledgeBaseDir: string, private manifest: Manifest, private agent?: PiAgentClient) {}

  async rewrite(query: string, targetDirs: string[] = [], n = 5, retryLevel = 1): Promise<string[]> {
    if (this.agent?.available) {
      const rewritten = await this.piRewrite(query, targetDirs, n, retryLevel);
      if (rewritten.length) return rewritten;
    }
    return this.heuristicRewrite(query, targetDirs, n, retryLevel);
  }

  private heuristicRewrite(query: string, targetDirs: string[], n: number, retryLevel: number): string[] {
    const terms = tokenize(query);
    const queries = [query];
    if (terms.length) queries.push(terms.join(" "));
    const contextTerms = this.contextTerms(query, targetDirs, retryLevel < 3 ? 12 : 20);
    if (contextTerms.length) {
      queries.push([...new Set([...terms, ...contextTerms.slice(0, 6)])].join(" "));
      queries.push(...contextTerms.slice(0, 3));
    }
    if (retryLevel >= 2) queries.push(...terms);
    if (retryLevel >= 3) queries.push(...contextTerms.slice(3, 8));
    if (retryLevel >= 4) queries.push(terms.slice(0, Math.max(1, Math.floor(terms.length / 2))).join(" "));
    return unique(queries).slice(0, n);
  }

  private async piRewrite(query: string, targetDirs: string[], n: number, retryLevel: number): Promise<string[]> {
    const result = await this.agent?.completeJson<string[]>(
      "Rewrite a question into concise keyword search queries for ripgrep over Markdown. Return only a JSON array of strings. Include the original query first.",
      JSON.stringify({
        query,
        targetDirs,
        retryLevel,
        maxQueries: n,
        manifestContext: this.manifestContext(targetDirs),
        indexContext: this.indexText(targetDirs).slice(0, 6000)
      })
    );
    if (!Array.isArray(result)) return [];
    return unique([query, ...result.map(String)]).slice(0, n);
  }

  private contextTerms(query: string, targetDirs: string[], limit: number): string[] {
    const queryTerms = new Set(tokenize(query));
    const ranked = this.manifestContext(targetDirs)
      .map((doc) => [scoreText(queryTerms, [doc.title, doc.summary, ...(doc.tags ?? [])].join(" ")), [doc.title, doc.summary, ...(doc.tags ?? [])].join(" ")] as const)
      .sort((a, b) => b[0] - a[0]);
    const sourceText = `${ranked.slice(0, 8).map(([, text]) => text).join(" ")} ${this.indexText(targetDirs)}`;
    return topTerms(sourceText, limit * 2).filter((term) => !queryTerms.has(term)).slice(0, limit);
  }

  private manifestContext(targetDirs: string[]): Array<{ path: string; title: string; summary: string; tags: string[]; categories: string[] }> {
    const targetSet = new Set(targetDirs);
    return this.manifest.documents
      .filter((doc) => {
        const firstPart = doc.path.split(/[\\/]/)[0];
        const categories = allCategories(doc);
        return !targetSet.size || targetSet.has(firstPart ?? "") || [...categories].some((category) => targetSet.has(category));
      })
      .slice(0, 80)
      .map((doc) => ({ path: doc.path, title: doc.title, summary: doc.summary, tags: doc.tags, categories: [...allCategories(doc)] }));
  }

  private indexText(targetDirs: string[]): string {
    return targetDirs.map((dir) => safeRead(path.join(this.knowledgeBaseDir, dir, "local_index.md"))).join("\n");
  }
}

function unique(values: string[]): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const value of values.map((item) => item.trim()).filter(Boolean)) {
    const key = value.toLowerCase();
    if (!seen.has(key)) {
      seen.add(key);
      out.push(value);
    }
  }
  return out;
}

function safeRead(file: string): string {
  try {
    return fs.readFileSync(file, "utf8");
  } catch {
    return "";
  }
}

File: app/agents/query_rewriter.py
from __future__ import annotations

from pathlib import Path

from app.schemas.document import Manifest
from app.text_utils import score_text, tokenize, top_terms


class QueryRewriter:
    def __init__(self, knowledge_base_dir: Path | None = None, manifest: Manifest | None = None):
        self.knowledge_base_dir = knowledge_base_dir
        self.manifest = manifest or Manifest()

    def rewrite(
        self,
        query: str,
        target_dirs: list[str] | None = None,
        n: int = 5,
        retry_level: int = 1,
    ) -> list[str]:
        terms = tokenize(query)
        queries = [query]
        if terms:
            queries.append(" ".join(terms))

        context_terms = self._context_terms(query, target_dirs or [], limit=12 if retry_level < 3 else 20)
        if context_terms:
            queries.append(" ".join(dict.fromkeys([*terms, *context_terms[:6]])))
            queries.extend(context_terms[:3])

        if retry_level >= 2 and terms:
            queries.extend(terms)
        if retry_level >= 3 and context_terms:
            queries.extend(context_terms[3:8])
        if retry_level >= 4:
            queries.append(" ".join(terms[: max(1, len(terms) // 2)]))
        return self._unique(queries)[:n]

    def _context_terms(self, query: str, target_dirs: list[str], limit: int) -> list[str]:
        query_terms = set(tokenize(query))
        ranked_docs = []
        target_set = set(target_dirs)
        for doc in self.manifest.documents:
            path_parts = Path(doc.path).parts
            in_target = bool(target_set & doc.all_categories) or bool(path_parts and path_parts[0] in target_set)
            if target_set and not in_target:
                continue
            text = " ".join(
                [
                    doc.title,
                    doc.summary,
                    *doc.tags,
                    *doc.departments,
                    doc.doc_type,
                    doc.primary_category,
                    *doc.secondary_categories,
                ]
            )
            ranked_docs.append((score_text(query_terms, text), text))

        ranked_docs.sort(key=lambda item: item[0], reverse=True)
        source_text = " ".join(text for _, text in ranked_docs[:8])
        source_text = " ".join([source_text, self._index_text(target_dirs)])
        terms = [term for term in top_terms(source_text, limit=limit * 2) if term not in query_terms]
        return terms[:limit]

    def _index_text(self, target_dirs: list[str]) -> str:
        if not self.knowledge_base_dir:
            return ""
        chunks = []
        for directory in target_dirs:
            local_index = self.knowledge_base_dir / directory / "local_index.md"
            if local_index.exists():
                chunks.append(local_index.read_text(encoding="utf-8", errors="replace"))
        return "\n".join(chunks)

    def _unique(self, values: list[str]) -> list[str]:
        seen = set()
        out = []
        for value in values:
            normalized = " ".join(value.split())
            if normalized and normalized.lower() not in seen:
                seen.add(normalized.lower())
                out.append(normalized)
        return out
